agent/auto-config/auto_config.go

```diff
@@ -498,3 +434,164 @@ func (ac *AutoConfig) getInitialConfiguration(ctx context.Context) error {
+
+	caConfig, err := ac.config.ConnectCAConfiguration()
+	if err != nil {
+		return "", "", fmt.Errorf("Cannot generate CSR: %w", err)
+	}
+
+	conf, err := caConfig.GetCommonConfig()
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to load common CA configuration: %w", err)
+	}
+
+	if conf.PrivateKeyType == "" {
+		conf.PrivateKeyType = connect.DefaultPrivateKeyType
+	}
+	if conf.PrivateKeyBits == 0 {
+		conf.PrivateKeyBits = connect.DefaultPrivateKeyBits
+	}
+
+	// Create a new private key
+	pk, pkPEM, err := connect.GeneratePrivateKeyWithConfig(conf.PrivateKeyType, conf.PrivateKeyBits)
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to generate private key: %w", err)
+	}
+
+	dnsNames := []string{"localhost"}
+	ipAddresses := []net.IP{net.ParseIP("127.0.0.1"), net.ParseIP("::")}
+
+	// Create a CSR.
+	//
+	// The Common Name includes the dummy trust domain for now but Server will
+	// override this when it is signed anyway so it's OK.
+	cn := connect.AgentCN(ac.config.NodeName, dummyTrustDomain)
+	csr, err = connect.CreateCSR(id, cn, pk, dnsNames, ipAddresses)
+	if err != nil {
+		return "", "", err
+	}
+
+	return csr, pkPEM, nil
+}
+
+// update will take an AutoConfigResponse and do all things necessary
+// to restore those settings. This currently involves updating the
+// config data to be used during a call to ReadConfig, updating the
+// tls Configurator and prepopulating the cache.
+func (ac *AutoConfig) update(resp *pbautoconf.AutoConfigResponse) error {
+	if err := ac.updateConfigFromResponse(resp); err != nil {
+		return err
+	}
+
+	if err := ac.updateTLSFromResponse(resp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// updateConfigFromResponse is responsible for generating the JSON compatible with the
+// agent/config.Config struct
+func (ac *AutoConfig) updateConfigFromResponse(resp *pbautoconf.AutoConfigResponse) error {
+	// here we want to serialize the translated configuration for use in injecting into the normal
+	// configuration parsing chain.
+	conf, err := json.Marshal(translateConfig(resp.Config))
+	if err != nil {
+		return fmt.Errorf("failed to encode auto-config configuration as JSON: %w", err)
+	}
+
+	ac.autoConfigData = string(conf)
+	return nil
+}
+
+// updateTLSFromResponse will update the TLS certificate and roots in the shared
+// TLS configurator.
+func (ac *AutoConfig) updateTLSFromResponse(resp *pbautoconf.AutoConfigResponse) error {
+	if ac.certMonitor == nil {
+		return nil
+	}
+
+	roots, err := translateCARootsToStructs(resp.CARoots)
+	if err != nil {
+		return err
+	}
+
+	cert, err := translateIssuedCertToStructs(resp.Certificate)
+	if err != nil {
+		return err
+	}
+
+	update := &structs.SignedResponse{
+		IssuedCert:     *cert,
+		ConnectCARoots: *roots,
+		ManualCARoots:  resp.ExtraCACertificates,
+	}
+
+	if resp.Config != nil && resp.Config.TLS != nil {
+		update.VerifyServerHostname = resp.Config.TLS.VerifyServerHostname
+	}
+
+	if err := ac.certMonitor.Update(update); err != nil {
```

Comment:
I was wondering if this also updates the TLSConfigurator - and it does!

Reply:
All of the logic of update the TLS configurator and managing the cache watches is internal to the `CertMonitor`. Update will prepopulate the cache and update the TLS configurator.